---
title: "Integrate old and new datasets"
author:
  - "trnguyen@ukaachen.de"
date: "Last update on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    keep_md: no
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: false
    toc_depth: 3
    number_sections: true
    theme: lumen
params:
  run: NA
  sample.id: NA
---


```{css zoom-lib-src, echo = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"
```

```{js zoom-jquery, echo = FALSE}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

<style type="text/css">
    div.datatables { height: auto !important;}
</style>

```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.height=10, fig.width=14}
##### clean up #####
gc()
rm(list = ls())

path.to.project.src <- "/home/hieunguyen/CRC1382/src_2023/CRC1382_AGBerres"
library(ggpubr)
source(file.path(path.to.project.src, "00_import_libraries.R"))
source(file.path(path.to.project.src, "00_helper_functions.R"))

path.to.src <- "/home/hieunguyen/CRC1382/src_2023/src_pipeline/scRNA_GEX_pipeline/processes_src"
source(file.path(path.to.src, "s8_integration_and_clustering.R"))

#####----------------------------------------------------------------------#####
# CONFIGURATIONS AND PREPRATIONS
#####----------------------------------------------------------------------#####
outdir <- "/home/hieunguyen/CRC1382/outdir"
PROJECT <- "ABeckers_231205"

path.to.main.input <- file.path(outdir, PROJECT)

path.to.main.output <- file.path(path.to.main.input, "data_analysis")
path.to.01.output <- file.path(path.to.main.output, "01_output")
path.to.03.output <- file.path(path.to.main.output, "03_output")
dir.create(path.to.03.output, showWarnings = FALSE, recursive = TRUE)

s.obj.new <- readRDS(file.path(path.to.01.output, "Sample_PV73_LV73.rds"))
s.obj.old <- readRDS(file.path("/media/hieunguyen/HD0/outdir/CRC1382/AGBerres/220311_Berres_MiedIII_scCITEseq/data_analysis/01_output/s_obj.annotated.integrated.tsne.rds"))
```

# UMAP of 2 datasets

## New dataset
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14}
DimPlot(object = s.obj.new, reduction = "INTE_TSNE", label = TRUE, label.box = TRUE, repel = TRUE)
```

## Previous dataset

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14}
DimPlot(object = s.obj.old, reduction = "INTE_TSNE", label = TRUE, label.box = TRUE, repel = TRUE)
```

# Integration of the two datasets

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14}
num.PCA <- 30
num.PC.used.in.UMAP <- 30
num.PC.used.in.Clustering <- 30
num.dim.integration <- 30
num.dim.cluster <- 30
cluster.resolution <- 0.5
chosen.seed <- 42
if (file.exists(file.path(path.to.03.output, "merge_all_6_samples.RNA_assay.rds")) == FALSE){
  chosen.assay <- "RNA"
  DefaultAssay(s.obj.old) <- chosen.assay
  DefaultAssay(s.obj.new) <- chosen.assay
  
  s.obj.old <- RunPCA(s.obj.old, npcs = num.PCA, verbose = FALSE, reduction.name="RNA_PCA")
  s.obj.new <- RunPCA(s.obj.new, npcs = num.PCA, verbose = FALSE, reduction.name="RNA_PCA")
  
  integration.anchors <- FindTransferAnchors(reference = s.obj.old, query = s.obj.new,
                                               dims = 1:25, reference.reduction = "RNA_PCA")
    
  ##### Transfer label from anchors to query dataset
  predictions <- TransferData(anchorset = integration.anchors, refdata = s.obj.old$cell.annotation,
                                dims = 1:25)
    
  s.obj.new <- AddMetaData(s.obj.new, metadata = predictions$predicted.id, col.name = "cell.annotation")

  s.obj <- merge(x = s.obj.old, 
                 y = s.obj.new,
                 merge.data = FALSE)

  chosen.assay <- "RNA"
  DefaultAssay(s.obj) <- chosen.assay
  
  s.obj <- NormalizeData(s.obj) # ---> use Log Normalized
  s.obj <- FindVariableFeatures(s.obj, selection.method = "vst")
  s.obj <- ScaleData(s.obj, features = rownames(s.obj))
  
  s.obj <- RunPCA(s.obj, npcs = num.PCA, verbose = FALSE, reduction.name=sprintf("%s_PCA", chosen.assay))
  s.obj <- RunUMAP(s.obj, reduction = sprintf("%s_PCA", chosen.assay), dims = 1:num.PCA, 
                             reduction.name=sprintf("%s_UMAP", chosen.assay), seed.use = chosen.seed)
  s.obj <- RunTSNE(s.obj, reduction = sprintf("%s_PCA", chosen.assay), dims = 1:num.PCA, 
                             reduction.name=sprintf("%s_TSNE", chosen.assay), seed.use = chosen.seed)
  
  s.obj <- FindNeighbors(s.obj, reduction = sprintf("%s_PCA", chosen.assay), dims = 1:num.PC.used.in.Clustering)
  
  s.obj <- FindClusters(s.obj, resolution = cluster.resolution, random.seed = chosen.seed)
  
  saveRDS(object = s.obj, file = file.path(path.to.03.output, "merge_all_6_samples.RNA_assay.rds"))  
  print("Finished saving the file")
  write.csv(data.frame(data = c("finished saving the object to disk")), file.path(path.to.03.output, "CHECK_merge_all_6_samples.RNA_assay.rds.csv"))
} else {
  s.obj <- readRDS(file.path(path.to.03.output, "merge_all_6_samples.RNA_assay.rds"))
}
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14}
source("/home/hieunguyen/CRC1382/src_2023/src_pipeline/scRNA_GEX_pipeline/processes_src/s8_integration_and_clustering.R")
if (file.exists(file.path(path.to.03.output, "s8_output", "merge_all_6_samples_test_function.output.s8.rds")) == FALSE){
  s.obj <- s8.integration.and.clustering(s.obj = s.obj, 
                           path.to.output = path.to.03.output, 
                           save.RDS.s8 = TRUE,
                           PROJECT = "merge_all_6_samples_test_function", 
                           num.dim.integration = num.dim.integration,
                           num.PCA = num.PCA,
                           num.PC.used.in.UMAP = num.PC.used.in.UMAP,
                           num.PC.used.in.Clustering = num.PC.used.in.Clustering,
                           cluster.resolution = cluster.resolution,
                           my_random_seed = 42,
                           umap.method = "uwot",
                           genes.to.not.run.PCA = NULL,
                           inte_pca_reduction_name = "INTE_PCA", 
                           inte_umap_reduction_name = "INTE_UMAP",
                           with.TSNE = TRUE,
                           k.filter = 200)  
} else {
  s.obj <- readRDS(file.path(path.to.03.output, "s8_output", "merge_all_6_samples_test_function.output.s8.rds"))
}

```

## TSNE: all clusters 
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=12}
DimPlot(object = s.obj, reduction = "INTE_TSNE", label = TRUE, label.box = TRUE, label.size = 10, pt.size = 1, repel = TRUE) + 
  ggtitle(sprintf("TSNE: All clusters")) +
  theme(axis.text = element_text(size=20, face = "bold"),
        axis.title = element_text(size=20, face = "bold"), 
        title = element_text(size = 20, face = "bold"))
```

## TSNE: all samples
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=12}
DimPlot(object = s.obj, reduction = "INTE_TSNE", label = TRUE, label.box = TRUE, label.size = 10, pt.size = 1, repel = TRUE, group.by = "name") + 
  ggtitle(sprintf("TSNE Sample: All samples after integrating")) +
  theme(axis.text = element_text(size=20, face = "bold"),
        axis.title = element_text(size=20, face = "bold"), 
        title = element_text(size = 20, face = "bold"))
```

### All 6 samples with cell annotation
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14}
DimPlot(object = s.obj, reduction = "INTE_TSNE", label = TRUE, label.box = TRUE, label.size = 10, pt.size = 1, repel = TRUE, group.by = "cell.annotation") + 
  ggtitle(sprintf("TSNE Sample: All samples after integrating, cell annotation")) +
  theme(axis.text = element_text(size=20, face = "bold"),
        axis.title = element_text(size=20, face = "bold"), 
        title = element_text(size = 20, face = "bold"))
```

### All old samples with cell annotation
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14}
DimPlot(object = subset(s.obj, name %in% c("LV73", "PV73") == FALSE), reduction = "INTE_TSNE", label = TRUE, label.box = TRUE, label.size = 10, pt.size = 1, repel = TRUE, group.by = "cell.annotation") + 
  ggtitle(sprintf("TSNE Sample: All samples after integrating, cell annotation")) +
  theme(axis.text = element_text(size=20, face = "bold"),
        axis.title = element_text(size=20, face = "bold"), 
        title = element_text(size = 20, face = "bold"))
```

### All new samples with cell annotation
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14}
DimPlot(object = subset(s.obj, name %in% c("LV73", "PV73") == TRUE), reduction = "INTE_TSNE", label = TRUE, label.box = TRUE, label.size = 10, pt.size = 1, repel = TRUE, group.by = "cell.annotation") + 
  ggtitle(sprintf("TSNE Sample: All samples after integrating, cell annotation")) +
  theme(axis.text = element_text(size=20, face = "bold"),
        axis.title = element_text(size=20, face = "bold"), 
        title = element_text(size = 20, face = "bold"))
```

# TSNE with highlighted DC1, DC2, DC3 and NCM, the rest in gray
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
Idents(s.obj) <- "cell.annotation"
DC1_cells <- WhichCells(object = s.obj, ident = "DC1")
DC2_cells <- WhichCells(object = s.obj, ident = "DC2")
DC3_cells <- WhichCells(object = s.obj, ident = "DC3")
NCM_cells <- WhichCells(object = s.obj, ident = "NCM")

tsne_dimplot_with_celltype.gray <- DimPlot(s.obj, reduction = "INTE_TSNE", cells.highlight= list(DC1_cells, DC2_cells, DC3_cells, NCM_cells), cols= "lightgray",
        label = TRUE, label.box = TRUE, repel = TRUE, pt.size = 1,
        cols.highlight = c("#f59089", "#7af5b4", "#6fa6ed", "#d0db39")) + theme(legend.position = "none")
tsne_dimplot_with_celltype.gray
```


# Cluster marker genes

Identify differentially expressed genes in each cluster. 
```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.height=10, fig.width=18}
if (file.exists(file.path(path.to.03.output, "DE_cluster_marker_genes.rds")) == FALSE){
  cluster.markers <- FindAllMarkers(object = s.obj, assay = "RNA", test.use = "wilcox")
  cluster.markers <- subset(cluster.markers, cluster.markers$p_val_adj < 0.05 & cluster.markers$avg_log2FC > 0)
  saveRDS(cluster.markers, file.path(path.to.03.output, "DE_cluster_marker_genes.rds"))
} else {
  cluster.markers <- readRDS(file.path(path.to.03.output, "DE_cluster_marker_genes.rds"))
}

```

## Feature plot {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (cluster.id in sort(unique(s.obj@meta.data$seurat_clusters))){
  DefaultAssay(s.obj) <- "RNA"
  cat(sprintf("### Cluster %s \n", cluster.id))
  tmp.cluster.markers <- subset(cluster.markers, cluster.markers$cluster == cluster.id) %>% arrange(desc(avg_log2FC))
  p <- FeaturePlot(object = s.obj, reduction = "INTE_TSNE", features = head(tmp.cluster.markers, 12)$gene, ncol = 3, label = TRUE, pt.size = 0.5, label.size = 5, label.color = "red")  
  print(p)
  cat("\n \n")
}
```

## Dot plot{.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (cluster.id in sort(unique(s.obj@meta.data$seurat_clusters))){
  DefaultAssay(s.obj) <- "RNA"
  cat(sprintf("### Cluster %s \n", cluster.id))
  tmp.cluster.markers <- subset(cluster.markers, cluster.markers$cluster == cluster.id) %>% arrange(desc(avg_log2FC))
  p <- DotPlot(object = s.obj, features = head(tmp.cluster.markers, 12)$gene)  
  print(p)
  cat("\n \n")
}
```

## Violin plot{.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (cluster.id in sort(unique(s.obj@meta.data$seurat_clusters))){
  DefaultAssay(s.obj) <- "RNA"
  cat(sprintf("### Cluster %s \n", cluster.id))
  tmp.cluster.markers <- subset(cluster.markers, cluster.markers$cluster == cluster.id) %>% arrange(desc(avg_log2FC))
  p <- VlnPlot(object = s.obj, features = head(tmp.cluster.markers, 12)$gene)  
  print(p)
  cat("\n \n")
}
```

## Full tables of DE genes {.tabset}
```{r echo=FALSE, results='asis', include=FALSE}
cluster.markers %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
all.cluster.ids <- sort(unique(s.obj$seurat_clusters))
for (cluster.id in all.cluster.ids){
  tmp.table <- subset(cluster.markers, cluster.markers$cluster == cluster.id) %>% arrange(desc(avg_log2FC))
cat(paste("\n\n### Cluster: ", cluster.id, "##\n"))
  print( htmltools::tagList(datatable(tmp.table, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))))
  cat("\n \n")  
}
```



# Cell type marker genes

Identify differentially expressed genes in each cell type 
```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.height=10, fig.width=18}
if (file.exists(file.path(path.to.03.output, "DE_celltype_marker_genes.rds")) == FALSE){
  Idents(s.obj) <- "cell.annotation"
  celltype.markers <- FindAllMarkers(object = s.obj, assay = "RNA", test.use = "wilcox")
  celltype.markers <- subset(celltype.markers, celltype.markers$p_val_adj < 0.05 & celltype.markers$avg_log2FC > 0)
  saveRDS(celltype.markers, file.path(path.to.03.output, "DE_celltype_marker_genes.rds"))
} else {
  celltype.markers <- readRDS(file.path(path.to.03.output, "DE_celltype_marker_genes.rds"))
}

```

## Feature plot {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (celltype.id in sort(unique(s.obj@meta.data$cell.annotation))){
  DefaultAssay(s.obj) <- "RNA"
  cat(sprintf("### celltype %s \n", celltype.id))
  tmp.celltype.markers <- subset(celltype.markers, celltype.markers$cluster == celltype.id) %>% arrange(desc(avg_log2FC))
  p <- FeaturePlot(object = s.obj, reduction = "INTE_TSNE", features = head(tmp.celltype.markers, 12)$gene, ncol = 3, label = TRUE, pt.size = 0.5, label.size = 5, label.color = "red")  
  print(p)
  cat("\n \n")
}
```

## Dot plot{.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (celltype.id in sort(unique(s.obj@meta.data$cell.annotation))){
  DefaultAssay(s.obj) <- "RNA"
  cat(sprintf("### celltype %s \n", celltype.id))
  tmp.celltype.markers <- subset(celltype.markers, celltype.markers$cluster == celltype.id) %>% arrange(desc(avg_log2FC))
  p <- DotPlot(object = s.obj, features = head(tmp.celltype.markers, 12)$gene)  
  print(p)
  cat("\n \n")
}
```

## Violin plot{.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (celltype.id in sort(unique(s.obj@meta.data$cell.annotation))){
  DefaultAssay(s.obj) <- "RNA"
  cat(sprintf("### celltype %s \n", celltype.id))
  tmp.celltype.markers <- subset(celltype.markers, celltype.markers$cluster == celltype.id) %>% arrange(desc(avg_log2FC))
  p <- VlnPlot(object = s.obj, features = head(tmp.celltype.markers, 12)$gene)  
  print(p)
  cat("\n \n")
}
```

## Full tables of DE genes {.tabset}
```{r echo=FALSE, results='asis', include=FALSE}
celltype.markers %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
all.celltype.ids <- sort(unique(s.obj$cell.annotation))
for (celltype.id in all.celltype.ids){
  tmp.table <- subset(celltype.markers, celltype.markers$cluster == celltype.id) %>% arrange(desc(avg_log2FC))
  cat(paste("\n\n### celltype: ", celltype.id, "##\n"))
  print( htmltools::tagList(datatable(tmp.table, extensions = 'Buttons',
                                      filter = "top",
                                      options = list(dom = 'Blfrtip',
                                                     buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                                                     lengthMenu = list(c(10,25,50,-1),
                                                                       c(10,25,50,"All")),
                                                     columnDefs = list(list(
                                                       targets = "_all",
                                                       render = JS(
                                                         "function(data, type, row, meta) {",
                                                         "return type === 'display' && data != null && data.length > 100 ?",
                                                         "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                                         "}")
                                                     ))
                                      ))))
  cat("\n \n")  
}
```

# Hashtag-Antibodies data (ADT)
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
DefaultAssay(s.obj) <- "ADT"

counts.adt <- GetAssayData(s.obj, slot = "counts", assay = "ADT")
s.obj.HTO <- CreateSeuratObject(counts = counts.adt, assay = "HTO")

s.obj.HTO <- NormalizeData(s.obj.HTO, normalization.method = "CLR")

```

## Histogram: Raw count number of hashtags per cell {.tabset}
Histogram: Raw count number of hashtags per cell: 

- x-axis: Number of hashtags per cell

- y-axis: Number of cells

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (ht in unique(row.names(s.obj.HTO))){
  hashtag.counts <- counts.adt[ht, ] %>% as.data.frame()
  colnames(hashtag.counts) <- c("count")
  cat(sprintf("### Hashtag: %s \n", ht))
  p <- ggplot(subset(hashtag.counts, hashtag.counts$count >= 0), aes(x = count)) + geom_histogram(bins = max(hashtag.counts$count)) + ggtitle(ht) + 
    theme(text = element_text(size=20), 
          axis.text.x = element_text(hjust=1, size = 20),
          axis.text.y = element_text(hjust=1, size = 20)) 
  print(p)
  cat("\n \n")
}
```

## Histogram: log-transformed raw counts of hashtags per cell {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (ht in  unique(row.names(s.obj.HTO))){
  counts.adt.ht123 <- GetAssayData(s.obj.HTO, slot = "counts")
  hashtag.counts <- log2(counts.adt.ht123)[ht, ] %>% as.data.frame()
  colnames(hashtag.counts) <- c("count")
  cat(sprintf("### Hashtag: %s \n", ht))
  p <- ggplot(subset(hashtag.counts, hashtag.counts$count >= 0), aes(x = count)) + geom_histogram(bins = 10) + 
    ggtitle(ht)  + 
    theme(text = element_text(size=20), 
          axis.text.x = element_text(hjust=1, size = 20),
          axis.text.y = element_text(hjust=1, size = 20)) 
  print(p)
  cat("\n \n")
}
```

## Assigned cut-off 
```{r echo=FALSE, warning=FALSE, results='hide', message=FALSE}
# s.obj.HTO <- HTODemux(s.obj.HTO, assay = "HTO", positive.quantile = 0.99)
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
# Idents(s.obj.HTO) <- factor(x = Idents(s.obj), levels = levels(s.obj))
```

## Illustration by Ridge plots, Hashtag-Antibodies expression values in each cluster {.tabset}

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=12, fig.height=12}
main.metadata <- s.obj@meta.data %>% rownames_to_column("barcode")
meta.data <- s.obj.HTO@meta.data %>% rownames_to_column("barcode")
meta.data <- merge(meta.data, main.metadata, by.x = "barcode", by.y = "barcode") %>%
  column_to_rownames("barcode")
meta.data <- meta.data[row.names(s.obj@meta.data),]
s.obj.HTO <- AddMetaData(object = s.obj.HTO, metadata = meta.data$cell.annotation, col.name = "cell.annotation")
Idents(s.obj.HTO) <- "cell.annotation"
for (hashtag in rownames(s.obj.HTO[["HTO"]])){
  cat(sprintf("### %s \n", hashtag))
  p <- RidgePlot(s.obj.HTO, assay = "HTO", features = hashtag) +
    theme(legend.position = "none")
  print(p)
  cat("\n \n")
}
```

## Illustration by Violin plot {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=12, fig.height=12}

for (hashtag in rownames(s.obj.HTO[["HTO"]])){
  cat(sprintf("### %s \n", hashtag))
  p <- VlnPlot(s.obj.HTO, assay = "HTO", features = c(hashtag)) + theme(legend.position = "none")
  print(p)
  cat("\n \n")
}
```

# Heatmaps

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
DefaultAssay(s.obj) <- "RNA"
s.obj.LV <- subset(s.obj, stage == "LV")
s.obj.PV <- subset(s.obj, stage == "PV")

cluster_of_interest <- c("DC1", "DC2", "DC3", "NCM")

gene_of_interest <- hash()
```


```{r echo=TRUE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
gene_of_interest[["Maturation"]] <- c("CD40", "CD86", "RELB", "CD83")
gene_of_interest[["Regulatory"]] <- c("CD274", "CD200", "FAS", "ALDH1A2", "SOCS1", "SOCS2", "BIRC3", "MARCKSL1")
gene_of_interest[["TLRs_and_Adaptors"]] <- c("TICAM1", "TLR9", "TLR8", "TLR7", "TLR6", "TLR5", "TLR4", "TLR2", "TLR1")
gene_of_interest[["Migration"]] <- c("CCR7", "MYO1G", "CXCL16", "ADAM8", "ICAM1", "FSCN1", "MARCKS", "MARCKSL1")
```
## List of genes to be shown on the heatmap

- Maturation: "CD40", "CD86", "RELB", "CD83"

- Regulatory: "CD274", "CD200", "FAS", "ALDH1A2", "SOCS1", "SOCS2", "BIRC3", "MARCKSL1"

- TLRs and Adaptors: "TICAM1", "TLR9", "TLR8", "TLR7", "TLR6", "TLR5", "TLR4", "TLR2", "TLR1"

- Migration: "CCR7", "MYO1G", "CXCL16", "ADAM8", "ICAM1", "FSCN1", "MARCKS", "MARCKSL1"

## Internal comparison between clusters in each LV, PV 2-sample 

### LV samples {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
DefaultAssay(s.obj) <- "RNA"
for (group in names(gene_of_interest)){
  cat(sprintf("#### %s \n", group))
  lv.df <- data.frame(gene_of_interest[[group]])
  colnames(lv.df) <- c("Gene")
  for (chosen.cluster in cluster_of_interest){
    tmp.s.obj.lv <- subset(s.obj.LV, cells = WhichCells(object = s.obj.LV, ident = chosen.cluster))
    count.matrix <- as.data.frame(GetAssayData(tmp.s.obj.lv, slot = "data"))
    count.matrix <- count.matrix[gene_of_interest[[group]], ]
    countdf <- as.data.frame(rowMeans(count.matrix))
    colnames(countdf) <- c(chosen.cluster)
    countdf <- countdf %>% rownames_to_column("Gene")
    lv.df <- merge(lv.df, countdf, by.x = "Gene", by.y = "Gene")
  }
    lv.df <- lv.df %>% column_to_rownames("Gene")
    lv.df$total <- rowSums(lv.df)
    # lv.df <- subset(lv.df, lv.df$total != 0)
    lv.df <- subset(lv.df, select = -c(total))
    # input.to.heatmap.lv <- log2(lv.df/t(rowMeans(lv.df)) + 0.1)
    input.to.heatmap.lv <- (lv.df - t(rowMeans(lv.df)))/(rowSds(as.matrix(lv.df)))
     
    tmp.heatmap.lv <- input.to.heatmap.lv %>% 
      as.data.frame() %>%
      rownames_to_column("Gene") %>%
      pivot_longer(-c(Gene), names_to = "Cluster", values_to = "Expression") %>%
      mutate(Cluster= fct_relevel(Cluster,colnames(input.to.heatmap.lv))) %>%
      ggplot(aes(x=Cluster, y=Gene, fill=Expression)) + 
      geom_raster() + 
    scale_fill_gradient2(low="#1417f5", high="#f52314", guide="colorbar") +
      theme(text = element_text(size=25))
    print(tmp.heatmap.lv)
    cat("\n \n")
}
```

### PV samples {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
DefaultAssay(s.obj) <- "RNA"
for (group in names(gene_of_interest)){
  cat(sprintf("#### %s \n", group))
  ##### PREPARE HEATMAP DATA FOR PV SAMPLES
  pv.df <- data.frame(gene_of_interest[[group]])#
  colnames(pv.df) <- c("Gene")
  for (chosen.cluster in cluster_of_interest){
    tmp.s.obj.pv <- subset(s.obj.PV, cells = WhichCells(object = s.obj.PV, ident = chosen.cluster))
    count.matrix <- as.data.frame(GetAssayData(tmp.s.obj.pv, slot = "data"))
    count.matrix <- count.matrix[gene_of_interest[[group]], ]
    countdf <- as.data.frame(rowMeans(count.matrix))
    colnames(countdf) <- c(chosen.cluster)
    countdf <- countdf %>% rownames_to_column("Gene")
    pv.df <- merge(pv.df, countdf, by.x = "Gene", by.y = "Gene")
  }
  pv.df <- pv.df %>% column_to_rownames("Gene")
  pv.df$total <- rowSums(pv.df)
  # pv.df <- subset(pv.df, pv.df$total != 0)
  pv.df <- subset(pv.df, select = -c(total))
  # input.to.heatmap.pv <- log2(pv.df/t(rowMeans(pv.df)) + 0.1)
  input.to.heatmap.pv <- (pv.df - t(rowMeans(pv.df)))/(rowSds(as.matrix(pv.df)))
  
  tmp.heatmap.pv <- input.to.heatmap.pv %>% 
    as.data.frame() %>%
    rownames_to_column("Gene") %>%
    pivot_longer(-c(Gene), names_to = "Cluster", values_to = "Expression") %>%
    mutate(Cluster= fct_relevel(Cluster,colnames(input.to.heatmap.pv))) %>%
    ggplot(aes(x=Cluster, y=Gene, fill=Expression)) + 
    geom_raster() + 
    scale_fill_gradient2(low="#1417f5", high="#f52314", guide="colorbar") +
    theme(text = element_text(size=25))
  print(tmp.heatmap.pv)
  cat("\n \n")
}

```


## Cross condition comparisons on heatmaps {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
DefaultAssay(s.obj) <- "RNA"
for (group in names(gene_of_interest)){
  cat(sprintf("### %s \n", group))
  lv.df <- data.frame(gene_of_interest[[group]])
  colnames(lv.df) <- c("Gene")
  for (chosen.cluster in cluster_of_interest){
    tmp.s.obj.lv <- subset(s.obj.LV, cells = WhichCells(object = s.obj.LV, ident = chosen.cluster))
    count.matrix <- as.data.frame(GetAssayData(tmp.s.obj.lv, slot = "data"))
    count.matrix <- count.matrix[gene_of_interest[[group]], ]
    countdf <- as.data.frame(rowMeans(count.matrix))
    colnames(countdf) <- c(chosen.cluster)
    countdf <- countdf %>% rownames_to_column("Gene")
    lv.df <- merge(lv.df, countdf, by.x = "Gene", by.y = "Gene")
  }
    lv.df <- lv.df %>% column_to_rownames("Gene")
    lv.df$total <- rowSums(lv.df)
    # lv.df <- subset(lv.df, lv.df$total != 0)
    lv.df <- subset(lv.df, select = -c(total))
    
  pv.df <- data.frame(gene_of_interest[[group]])
  colnames(pv.df) <- c("Gene")
  for (chosen.cluster in cluster_of_interest){
    tmp.s.obj.pv <- subset(s.obj.PV, cells = WhichCells(object = s.obj.PV, ident = chosen.cluster))
    count.matrix <- as.data.frame(GetAssayData(tmp.s.obj.pv, slot = "data"))
    count.matrix <- count.matrix[gene_of_interest[[group]], ]
    countdf <- as.data.frame(rowMeans(count.matrix))
    colnames(countdf) <- c(chosen.cluster)
    countdf <- countdf %>% rownames_to_column("Gene")
    pv.df <- merge(pv.df, countdf, by.x = "Gene", by.y = "Gene")
    }
  pv.df <- pv.df %>% column_to_rownames("Gene")
  pv.df$total <- rowSums(pv.df)
  # pv.df <- subset(pv.df, pv.df$total != 0)
  pv.df <- subset(pv.df, select = -c(total))

  colnames(lv.df) <- to_vec(for (item in colnames(lv.df)) sprintf("LV_%s", item))
  colnames(pv.df) <- to_vec(for (item in colnames(pv.df)) sprintf("PV_%s", item))
  
  lv.df <- lv.df %>% rownames_to_column("Gene")
  pv.df <- pv.df %>% rownames_to_column("Gene")
  
  merge.lv.pv.df <- merge(pv.df, lv.df, by.x = "Gene", by.y = "Gene") %>%
    column_to_rownames("Gene")

  input.to.heatmap <- (merge.lv.pv.df - t(rowMeans(merge.lv.pv.df)))/(rowSds(as.matrix(merge.lv.pv.df)))
  tmp.heatmap.lv.pv <- input.to.heatmap %>% 
    as.data.frame() %>%
    rownames_to_column("Gene") %>%
    pivot_longer(-c(Gene), names_to = "Cluster", values_to = "Expression") %>%
    mutate(Cluster= fct_relevel(Cluster,colnames(input.to.heatmap))) %>%
    ggplot(aes(x=Cluster, y=Gene, fill=Expression)) + 
    geom_raster() + 
    scale_fill_gradient2(low="#1417f5", high="#f52314", guide="colorbar") +
        theme(text = element_text(size=25))
  
  input.to.heatmap.lv.pv.only.lv <- input.to.heatmap[, c("LV_DC1", "LV_DC2", "LV_DC3", "LV_NCM")]
  input.to.heatmap.lv.pv.only.pv <- input.to.heatmap[, c("PV_DC1", "PV_DC2", "PV_DC3", "PV_NCM")]
  
  tmp.heatmap.lv <- input.to.heatmap.lv.pv.only.lv %>% 
    as.data.frame() %>%
    rownames_to_column("Gene") %>%
    pivot_longer(-c(Gene), names_to = "Cluster", values_to = "Expression") %>%
    mutate(Cluster= fct_relevel(Cluster,colnames(input.to.heatmap.lv.pv.only.lv))) %>%
    ggplot(aes(x=Cluster, y=Gene, fill=Expression)) + 
    geom_raster() + 
    scale_fill_gradient2(low="#1417f5", high="#f52314", guide="colorbar") +
        theme(text = element_text(size=25))
  
  tmp.heatmap.pv <- input.to.heatmap.lv.pv.only.pv %>% 
    as.data.frame() %>%
    rownames_to_column("Gene") %>%
    pivot_longer(-c(Gene), names_to = "Cluster", values_to = "Expression") %>%
    mutate(Cluster= fct_relevel(Cluster,colnames(input.to.heatmap.lv.pv.only.pv))) %>%
    ggplot(aes(x=Cluster, y=Gene, fill=Expression)) + 
    geom_raster() + 
    scale_fill_gradient2(low="#1417f5", high="#f52314", guide="colorbar") +
        theme(text = element_text(size=25))
  print(tmp.heatmap.lv.pv)
  cat("\n \n")
  }
```

# Violin plot for Surface markers of DC1, DC2, DC3 and NCM cells {.tabset}

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
DefaultAssay(s.obj) <- "ADT"
s.obj <- ScaleData(s.obj, assay = "ADT")
s.obj <- NormalizeData(s.obj, assay = "ADT", normalization.method = "CLR")
Idents(s.obj) <- factor(x = Idents(s.obj), levels = sort(levels(s.obj)))

for (chosen.adt in row.names(s.obj)){
  cat(sprintf("## %s \n", chosen.adt))
  tmp.s.obj.violin <- s.obj[, Idents(s.obj) %in% c("DC1", "DC2", "DC3", "NCM")]
  tmp.s.obj.violin$stage <- factor(x = tmp.s.obj.violin$stage, levels = c("PV", "LV"))
  tmp.violin.plot <- VlnPlot(object = tmp.s.obj.violin, 
                             features = chosen.adt, pt.size = 0.05, assay = "ADT", split.by = "stage")
  print(tmp.violin.plot)
  cat("\n \n")
}
```


<!-- # Difference between 2 batches of data -->

<!-- ## Old batch vs new batch -->
<!-- ```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14} -->
<!-- meta.data <- s.obj@meta.data %>% -->
<!--   rownames_to_column("barcode") %>% -->
<!--   mutate(check.group = ifelse(name %in% c("LV20", "LV21", "PV20", "PV21"), "old_data", "new_data")) %>% -->
<!--   mutate(check.sample = sprintf("%s_%s", check.group, stage)) %>% -->
<!--   column_to_rownames("barcode") -->

<!-- meta.data <- meta.data[row.names(s.obj@meta.data),] -->

<!-- s.obj <- AddMetaData(object = s.obj, metadata = meta.data$check.group, col.name = "check.group") -->
<!-- s.obj <- AddMetaData(object = s.obj, metadata = meta.data$check.sample, col.name = "check.sample") -->
<!-- ``` -->


<!-- ```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14} -->
<!-- DimPlot(object = s.obj, reduction = "INTE_TSNE", group.by = "check.group", label = TRUE, label.box = TRUE, repel = TRUE) -->
<!-- ``` -->


<!-- ```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14} -->
<!-- DimPlot(object = s.obj, reduction = "INTE_TSNE", group.by = "check.sample", label = TRUE, label.box = TRUE, repel = TRUE) -->
<!-- ``` -->

<!-- ## Comparison of QC metrics and differentially expressed gene between 2 batches -->

<!-- ### QC metrics {.tabset} -->
<!-- ```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14} -->
<!-- meta.data <- s.obj@meta.data %>% rownames_to_column("barcode") -->
<!-- for (metric.to.plot in c("percent.mt", "percent.ribo", "nFeature_RNA", "nCount_RNA",  -->
<!--                          "AmbientRNA", "log10GenesPerUMI")){ -->
<!--   cat(sprintf("#### %s \n", metric.to.plot)) -->
<!--   p <- meta.data %>% ggplot(aes_string(x = "check.group", y = metric.to.plot, fill = "check.group")) +  -->
<!--     geom_boxplot() + theme_pubr() +  -->
<!--   stat_compare_means(method = "t.test", show.legend = TRUE, size = 12)   -->
<!--   print(p) -->
<!--   cat("\n \n") -->
<!-- } -->

<!-- ``` -->

<!-- ### Gene expression: All "old" samples vs all "new" samples -->
<!-- ```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14} -->
<!-- s.obj <- FindVariableFeatures(s.obj, assay = "RNA") -->
<!-- var.features <- VariableFeatures(s.obj) -->
<!-- ``` -->


<!-- ```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14} -->
<!-- if (file.exists(file.path(path.to.03.output, "compare_diff_genes.old_vs_new_batch.filtered.rds")) == FALSE){ -->
<!--   old.new.markers <- FindMarkers(object = s.obj, group.by = "check.group", ident.1 = "old_data", ident.2 = "new_data", assay = "RNA", test.use = "wilcox") -->
<!--   saveRDS(old.new.markers, file.path(path.to.03.output, "compare_diff_genes.old_vs_new_batch.raw.rds")) -->
<!--   old.new.markers <- subset(old.new.markers, old.new.markers$p_val_adj <= 0.05) -->
<!--   saveRDS(old.new.markers, file.path(path.to.03.output, "compare_diff_genes.old_vs_new_batch.filtered.rds"))   -->
<!-- } else { -->
<!--   old.new.markers <- readRDS(file.path(path.to.03.output, "compare_diff_genes.old_vs_new_batch.filtered.rds")) -->
<!-- } -->

<!-- old.new.markers <- old.new.markers %>%  -->
<!--   rownames_to_column("Gene") %>% -->
<!--   rowwise() %>% -->
<!--   mutate(abs_avgLog2FC = abs(avg_log2FC)) %>% -->
<!--   arrange(desc(abs_avgLog2FC)) %>% -->
<!--   mutate(used_in_TSNE = ifelse(Gene %in% var.features, "yes", "no")) %>% -->
<!--   mutate(gene_only_in_old_batch = ifelse(pct.1 >= 0.25 & pct.2 == 0, "yes", "no")) %>% -->
<!--   mutate(gene_only_in_new_batch = ifelse(pct.1 == 0 & pct.2 >= 0.25, "yes", "no")) -->
<!-- ``` -->

<!-- #### Table -->
<!-- ```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14} -->
<!-- old.new.markers %>% create_dt() -->
<!-- ``` -->



<!-- ### Gene expression:"Old" LV samples vs "new" LV samples -->
<!-- ```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14} -->
<!-- if (file.exists(file.path(path.to.03.output, "compare_diff_genes.old_vs_new_batch.LV_only.filtered.rds")) == FALSE){ -->
<!--   old.new.markers.LV <- FindMarkers(object = s.obj, group.by = "check.sample", ident.1 = "old_data_LV", ident.2 = "new_data_LV", assay = "RNA", test.use = "wilcox") -->
<!--   saveRDS(old.new.markers.LV, file.path(path.to.03.output, "compare_diff_genes.old_vs_new_batch.LV_only.raw.rds")) -->
<!--   old.new.markers.LV <- subset(old.new.markers.LV, old.new.markers.LV$p_val_adj <= 0.05) -->
<!--   saveRDS(old.new.markers.LV, file.path(path.to.03.output, "compare_diff_genes.old_vs_new_batch.LV_only.filtered.rds"))   -->
<!-- } else { -->
<!--   old.new.markers.LV <- readRDS(file.path(path.to.03.output, "compare_diff_genes.old_vs_new_batch.LV_only.filtered.rds")) -->
<!-- } -->
<!-- old.new.markers.LV <- old.new.markers.LV %>%  -->
<!--   rownames_to_column("Gene") %>% -->
<!--   rowwise() %>% -->
<!--   mutate(abs_avgLog2FC = abs(avg_log2FC)) %>% -->
<!--   arrange(desc(abs_avgLog2FC)) %>% -->
<!--   mutate(used_in_TSNE = ifelse(Gene %in% var.features, "yes", "no")) %>% -->
<!--   mutate(gene_only_in_old_batch = ifelse(pct.1 >= 0.25 & pct.2 == 0, "yes", "no")) %>% -->
<!--   mutate(gene_only_in_new_batch = ifelse(pct.1 == 0 & pct.2 >= 0.25, "yes", "no")) -->
<!-- ``` -->

<!-- #### Table -->
<!-- ```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14} -->
<!-- old.new.markers.LV %>% create_dt() -->
<!-- ``` -->

<!-- ### Gene expression:"Old" PV samples vs "new" PV samples -->
<!-- ```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14} -->
<!-- if (file.exists(file.path(path.to.03.output, "compare_diff_genes.old_vs_new_batch.PV_only.filtered.rds")) == FALSE){ -->
<!--   old.new.markers.PV <- FindMarkers(object = s.obj, group.by = "check.sample", ident.1 = "old_data_PV", ident.2 = "new_data_PV", assay = "RNA", test.use = "wilcox") -->
<!--   saveRDS(old.new.markers.PV, file.path(path.to.03.output, "compare_diff_genes.old_vs_new_batch.PV_only.raw.rds")) -->
<!--   old.new.markers.PV <- subset(old.new.markers.PV, old.new.markers.PV$p_val_adj <= 0.05) -->
<!--   saveRDS(old.new.markers.PV, file.path(path.to.03.output, "compare_diff_genes.old_vs_new_batch.PV_only.filtered.rds"))   -->
<!-- } else { -->
<!--   old.new.markers.PV <- readRDS(file.path(path.to.03.output, "compare_diff_genes.old_vs_new_batch.PV_only.filtered.rds")) -->
<!-- } -->

<!-- old.new.markers.PV <- old.new.markers.PV %>%  -->
<!--   rownames_to_column("Gene") %>% -->
<!--   rowwise() %>% -->
<!--   mutate(abs_avgLog2FC = abs(avg_log2FC)) %>% -->
<!--   arrange(desc(abs_avgLog2FC)) %>% -->
<!--   mutate(used_in_TSNE = ifelse(Gene %in% var.features, "yes", "no")) %>% -->
<!--   mutate(gene_only_in_old_batch = ifelse(pct.1 >= 0.25 & pct.2 == 0, "yes", "no")) %>% -->
<!--   mutate(gene_only_in_new_batch = ifelse(pct.1 == 0 & pct.2 >= 0.25, "yes", "no")) -->
<!-- ``` -->

<!-- #### Table -->
<!-- ```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14} -->
<!-- old.new.markers.PV %>% create_dt() -->
<!-- ``` -->

# Diffferential gene expression analysis

We perform differential gene expression analysis for each cell type between all LV samples vs all PV samples. 

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14}
if (file.exists(file.path(path.to.03.output, "diff_markers.rds")) == FALSE){
  diff.markers <- hash()
  raw.diff.markers <- hash()
  
  DefaultAssay(s.obj) <- "RNA"
  for (input.celltype in unique(s.obj$cell.annotation)){
    subset.s.obj <- subset(s.obj, cell.annotation == input.celltype)
    tmp <- FindMarkers(object = subset.s.obj, assay = "RNA", ident.1 = "LV", ident.2 = "PV", group.by = "stage")
    tmp.raw <- tmp %>% rownames_to_column("Gene")
    tmp <- tmp %>%
      rownames_to_column("Gene") %>%
      subset(p_val_adj <= 0.05) %>%
      rowwise() %>%
      mutate(abs_log2FC = abs(avg_log2FC)) %>%
      arrange(desc(abs_log2FC))
    diff.markers[[input.celltype]] <- tmp
    raw.diff.markers[[input.celltype]] <- tmp.raw
  }
  
  saveRDS(diff.markers, file.path(path.to.03.output, "diff_markers.rds"))
  saveRDS(raw.diff.markers, file.path(path.to.03.output, "raw_diff_markers.rds"))  
} else {
  diff.markers <- readRDS(file.path(path.to.03.output, "diff_markers.rds"))
  raw.diff.markers <- readRDS(file.path(path.to.03.output, "raw_diff_markers.rds"))
}

```

## Tables {.tabset}
```{r echo=FALSE, results='asis', include=FALSE}
diff.markers$DC1 %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}

for (cluster.id in names(diff.markers)){
  tmp.table <- diff.markers[[cluster.id]]
  cat(paste("\n\n### Cluster: ", cluster.id, "##\n"))
  print( htmltools::tagList(datatable(tmp.table, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))))
  cat("\n \n")  
}
```

## Dot plot, all samples grouped by condition {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (cluster.id in names(diff.markers)){
  DefaultAssay(s.obj) <- "RNA"
  cat(sprintf("### Cluster %s \n", cluster.id))
  tmp.cluster.markers <- diff.markers[[cluster.id]] 
  if (nrow(tmp.cluster.markers) != 0){
    tmp.cluster.markers <- tmp.cluster.markers %>% arrange(desc(abs_log2FC))
    Idents(s.obj) <- "stage"
    p <- DotPlot(object = subset(s.obj, cell.annotation == cluster.id), features = head(tmp.cluster.markers, 12)$Gene)  
  } else{
    p <- ggplot() + ggtitle("No differentially expressed gene to show.")
  }
  print(p)
  cat("\n \n")
}
```

## Dot plot, each sample {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (cluster.id in names(diff.markers)){
  DefaultAssay(s.obj) <- "RNA"
  cat(sprintf("### Cluster %s \n", cluster.id))
  tmp.cluster.markers <- diff.markers[[cluster.id]] 
  if (nrow(tmp.cluster.markers) != 0){
    tmp.cluster.markers <- tmp.cluster.markers %>% arrange(desc(abs_log2FC))
    Idents(s.obj) <- "name"
    p <- DotPlot(object = subset(s.obj, cell.annotation == cluster.id), features = head(tmp.cluster.markers, 12)$Gene)  
  } else{
    p <- ggplot() + ggtitle("No differentially expressed gene to show.")
  }
  print(p)
  cat("\n \n")
}
```

## Violin plot, all samples grouped by condition {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (cluster.id in names(diff.markers)){
  DefaultAssay(s.obj) <- "RNA"
  cat(sprintf("### Cluster %s \n", cluster.id))
  tmp.cluster.markers <- diff.markers[[cluster.id]] 
  if (nrow(tmp.cluster.markers) != 0){
    tmp.cluster.markers <- tmp.cluster.markers %>% arrange(desc(abs_log2FC))
    Idents(s.obj) <- "stage"
    p <- VlnPlot(object = subset(s.obj, cell.annotation == cluster.id), features = head(tmp.cluster.markers, 12)$Gene, pt.size = 0)  
  } else{
    p <- ggplot() + ggtitle("No differentially expressed gene to show.")
  }
  print(p)
  cat("\n \n")
}
```

## Violin plot, each sample {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (cluster.id in names(diff.markers)){
  DefaultAssay(s.obj) <- "RNA"
  cat(sprintf("### Cluster %s \n", cluster.id))
  tmp.cluster.markers <- diff.markers[[cluster.id]] 
  if (nrow(tmp.cluster.markers) != 0){
    tmp.cluster.markers <- tmp.cluster.markers %>% arrange(desc(abs_log2FC))
    Idents(s.obj) <- "name"
    p <- VlnPlot(object = subset(s.obj, cell.annotation == cluster.id), features = head(tmp.cluster.markers, 12)$Gene, pt.size = 0)  
  } else{
    p <- ggplot() + ggtitle("No differentially expressed gene to show.")
  }
  print(p)
  cat("\n \n")
}
```

## Volcano plots {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14}
for(cluster.id in names(diff.markers)){
  cat(sprintf("### %s \n", cluster.id))
  tmp <- raw.diff.markers[[cluster.id]] 
  tmp.filtered <- diff.markers[[cluster.id]] %>% arrange(desc(abs_log2FC))
  gene.to.plot <- head(tmp.filtered, 10)$Gene
  tmp <- tmp %>%
    rowwise() %>%
    mutate(sig = ifelse(p_val_adj <= 0.05, "significant", "not significant")) %>%
    mutate(volcano.Gene = ifelse(Gene %in% gene.to.plot, Gene, NA))
  
  volcano.plot <- ggplot(data=tmp, aes(x=avg_log2FC, y=-log10(p_val_adj), col=sig, label=volcano.Gene)) +
    geom_point() + 
    scale_color_manual(values=c("#c0d2f0", "#f28095")) +
    theme_minimal() +
    geom_vline(xintercept=c(0), col="#9a9fa6", linetype='dotted') +
    geom_vline(xintercept=c(-1, 1), col="#9a9fa6", linetype='dotted') +
    geom_hline(yintercept=-log10(0.05), col="#9a9fa6", linetype='dotted') +
    theme_bw() +
    theme(plot.title = element_text(hjust=0.5, face="bold", size = 12)) +
    geom_text_repel(max.overlaps = 10000,
                    segment.colour = "black", fontface = 'bold', seed = 42, colour = "red", size = 6)
  
  print(volcano.plot)  
  cat("\n \n")
}

```

<!-- # Differential gene expression analysis by pseudobulk-RNA data -->

<!-- ```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14} -->
<!-- library(DESeq2) -->

<!-- test.celltype <- "DC3" -->

<!-- test.barcodes <- hash() -->
<!-- for (sample.id in unique(s.obj$name)){ -->
<!--   test.barcodes[[sample.id]] <- subset(s.obj, cell.annotation == test.celltype & name == sample.id) %>% colnames()   -->
<!-- } -->

<!-- pseudobulk.matrix <- data.frame(Gene = row.names(s.obj)) -->
<!-- condition1 <- "LV" -->
<!-- condition2 <- "PV" -->

<!-- s.obj.list <- hash() -->
<!-- s.obj.list[[condition1]] <- subset(s.obj, stage == condition1) -->
<!-- s.obj.list[[condition2]] <- subset(s.obj, stage == condition2) -->

<!-- DefaultAssay(s.obj.list[[condition1]]) <- "RNA" -->
<!-- DefaultAssay(s.obj.list[[condition2]]) <- "RNA" -->

<!-- count.matrix <- hash() -->
<!-- for (condition in c(condition1, condition2)){ -->
<!--   count.matrix[[condition]] <- hash() -->
<!--   for (sample.id in unique(subset(s.obj, stage == condition)$name)){ -->
<!--     tmp.count.matrix <- GetAssayData(subset(s.obj, name == sample.id), slot = "count", assay = "RNA")[, test.barcodes[[sample.id]]] -->
<!--     rowsum.tmp <- rowSums(tmp.count.matrix)  -->
<!--     rowsum.tmp <- rowsum.tmp[rowsum.tmp >= 10] -->
<!--     tmp.count.matrix <- rowsum.tmp %>%  as.data.frame() %>% -->
<!--       rownames_to_column("Gene") -->
<!--     colnames(tmp.count.matrix) <- c("Gene", sprintf("%s", sample.id)) -->
<!--     count.matrix[[condition]][[sample.id]] <- tmp.count.matrix -->
<!--     pseudobulk.matrix <- merge(pseudobulk.matrix, tmp.count.matrix, by.x = "Gene", by.y = "Gene", all.x = FALSE) -->
<!--   } -->
<!-- } -->

<!-- pseudobulk.matrix <- pseudobulk.matrix %>% column_to_rownames("Gene") -->
<!-- pseudo_metadata <- data.frame(sample.id = colnames(pseudobulk.matrix)) %>% -->
<!--   rowwise() %>% -->
<!--   mutate(group = ifelse( grepl("LV", sample.id), "LV", "PV")) -->

<!-- pseudo_metadata$group <- factor(pseudo_metadata$group, levels = c(condition1, condition2)) -->

<!-- dds <- DESeqDataSetFromMatrix(pseudobulk.matrix,  -->
<!--                               colData = pseudo_metadata,  -->
<!--                               design = ~ group) -->

<!-- dds <- DESeq(dds) -->
<!-- contrast <- c("group", levels(pseudo_metadata$group)[1], levels(pseudo_metadata$group)[2]) -->

<!-- res <- results(dds,  -->
<!--                contrast = contrast, -->
<!--                alpha = 0.05) -->

<!-- resdf <- res %>% -->
<!--   data.frame() %>% rownames_to_column("Gene") -->

<!-- padj_cutoff <- 0.05 -->
<!-- logFC_cutoff <- 1 -->

<!-- normalized_counts <- counts(dds, normalized = TRUE) %>% as.data.frame() %>% rownames_to_column("Gene") -->
<!-- resdf <- merge(resdf, normalized_counts, by.x = "Gene", by.y = "Gene") %>% -->
<!--   mutate(abslog2FoldChange = abs(log2FoldChange)) %>% -->
<!--   mutate(sig = ifelse(padj <= padj_cutoff & abslog2FoldChange >= logFC_cutoff, "sig.diff", "nonsig.diff"))  -->

<!-- resdf.sigOnly <- subset(resdf, resdf$padj <= padj_cutoff & resdf$abslog2FoldChange >= logFC_cutoff) %>% arrange(desc(log2FoldChange)) -->
<!-- heatmap.input <- rbind(head(resdf.sigOnly, 10), tail(resdf.sigOnly, 10))[, c("Gene", pseudo_metadata$sample.id)] -->

<!-- heatmap.top50.input <- rbind(head(resdf.sigOnly, 50), tail(resdf.sigOnly, 50))[, c("Gene", pseudo_metadata$sample.id)] -->

<!-- heatmap.full.input <-resdf.sigOnly[, c("Gene", pseudo_metadata$sample.id)] %>% column_to_rownames("Gene") -->
<!-- heatmap.full.input.scaled <- (heatmap.full.input - rowSums(heatmap.full.input))/rowSds(as.matrix(heatmap.full.input))  -->

<!-- heatmap.full.input.scaled <- heatmap.full.input.scaled %>% rownames_to_column("Gene")  -->
<!-- heatmap.full.input.scaled <- merge(heatmap.full.input.scaled, subset(resdf.sigOnly, select = c(Gene, padj, log2FoldChange, abslog2FoldChange)), -->
<!--                                    by.x = "Gene", by.y = "Gene") %>% arrange(desc(log2FoldChange)) -->
<!-- ``` -->


# Diffferential surface marker expression analysis

We perform differential gene expression analysis for each cell type between all LV samples vs all PV samples. 

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14}
if (file.exists(file.path(path.to.03.output, "diff_surface_markers.rds")) == FALSE){
  diff.surface.markers <- hash()
  raw.diff.surface.markers <- hash()
  
  DefaultAssay(s.obj) <- "ADT"
  s.obj <- NormalizeData(s.obj)
  for (input.celltype in unique(s.obj$cell.annotation)){
    subset.s.obj <- subset(s.obj, cell.annotation == input.celltype)
    
    tmp <- FindMarkers(object = subset.s.obj, assay = "ADT", ident.1 = "LV", ident.2 = "PV", group.by = "stage", slot = "data", test.use = "wilcox")
    tmp.raw <- tmp %>% rownames_to_column("Gene")
    tmp <- tmp %>%
      rownames_to_column("Gene") %>%
      subset(p_val_adj <= 0.05) %>%
      rowwise() %>%
      mutate(abs_log2FC = abs(avg_log2FC)) %>%
      arrange(desc(abs_log2FC))
    diff.surface.markers[[input.celltype]] <- tmp
    raw.diff.surface.markers[[input.celltype]] <- tmp.raw
  }
  
  saveRDS(diff.surface.markers, file.path(path.to.03.output, "diff_surface_markers.rds"))
  saveRDS(raw.diff.surface.markers, file.path(path.to.03.output, "raw_diff_surface_markers.rds"))
} else {
  diff.surface.markers <- readRDS(file.path(path.to.03.output, "diff_surface_markers.rds"))
  raw.diff.surface.markers <- readRDS(file.path(path.to.03.output, "raw_diff_surface_markers.rds"))
}

```

## Tables {.tabset}
```{r echo=FALSE, results='asis', include=FALSE}
diff.surface.markers$DC1 %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}

for (cluster.id in names(diff.surface.markers)){
  tmp.table <- diff.surface.markers[[cluster.id]]
  cat(paste("\n\n### Cluster: ", cluster.id, "##\n"))
  print( htmltools::tagList(datatable(tmp.table, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))))
  cat("\n \n")  
}
```


## Dot plot, all samples grouped by condition {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (cluster.id in names(diff.surface.markers)){
  DefaultAssay(s.obj) <- "ADT"
  cat(sprintf("### Cluster %s \n", cluster.id))
  tmp.cluster.markers <- diff.surface.markers[[cluster.id]] 
  if (nrow(tmp.cluster.markers) != 0){
    tmp.cluster.markers <- tmp.cluster.markers %>% arrange(desc(abs_log2FC))
    Idents(s.obj) <- "stage"
    p <- DotPlot(object = subset(s.obj, cell.annotation == cluster.id), features = head(tmp.cluster.markers, 12)$Gene)  
  } else{
    p <- ggplot() + ggtitle("No differentially expressed gene to show.")
  }
  print(p)
  cat("\n \n")
}
```

## Dot plot, each sample {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (cluster.id in names(diff.surface.markers)){
  DefaultAssay(s.obj) <- "ADT"
  cat(sprintf("### Cluster %s \n", cluster.id))
  tmp.cluster.markers <- diff.surface.markers[[cluster.id]] 
  if (nrow(tmp.cluster.markers) != 0){
    tmp.cluster.markers <- tmp.cluster.markers %>% arrange(desc(abs_log2FC))
    Idents(s.obj) <- "name"
    p <- DotPlot(object = subset(s.obj, cell.annotation == cluster.id), features = head(tmp.cluster.markers, 12)$Gene)  
  } else{
    p <- ggplot() + ggtitle("No differentially expressed gene to show.")
  }
  print(p)
  cat("\n \n")
}
```

## Violin plot, all samples grouped by condition {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (cluster.id in names(diff.surface.markers)){
  DefaultAssay(s.obj) <- "ADT"
  cat(sprintf("### Cluster %s \n", cluster.id))
  tmp.cluster.markers <- diff.surface.markers[[cluster.id]] 
  if (nrow(tmp.cluster.markers) != 0){
    tmp.cluster.markers <- tmp.cluster.markers %>% arrange(desc(abs_log2FC))
    Idents(s.obj) <- "stage"
    p <- VlnPlot(object = subset(s.obj, cell.annotation == cluster.id), features = head(tmp.cluster.markers, 12)$Gene, pt.size = 0)  
  } else{
    p <- ggplot() + ggtitle("No differentially expressed gene to show.")
  }
  print(p)
  cat("\n \n")
}
```

## Violin plot, each sample {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for (cluster.id in names(diff.surface.markers)){
  DefaultAssay(s.obj) <- "ADT"
  cat(sprintf("### Cluster %s \n", cluster.id))
  tmp.cluster.markers <- diff.surface.markers[[cluster.id]] 
  if (nrow(tmp.cluster.markers) != 0){
    tmp.cluster.markers <- tmp.cluster.markers %>% arrange(desc(abs_log2FC))
    Idents(s.obj) <- "name"
    p <- VlnPlot(object = subset(s.obj, cell.annotation == cluster.id), features = head(tmp.cluster.markers, 12)$Gene, pt.size = 0)  
  } else{
    p <- ggplot() + ggtitle("No differentially expressed gene to show.")
  }
  print(p)
  cat("\n \n")
}
```


```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.height=10, fig.width=14}
```